core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -118,8 +126,22 @@
 
   private final class ContextCancellationListener implements CancellationListener {
     @Override
-    public void cancelled(Context context) {
-      stream.cancel(statusFromCancelled(context));
+    public void cancelled(final Context context) {
+      if (context.getDeadline() == null || !context.getDeadline().isExpired()) {
+        stream.cancel(statusFromCancelled(context));
+      } else {
+        if (deadlineCancellationSendToServerFuture == null) {
+          deadlineCancellationSendToServerFuture = deadlineCancellationExecutor.schedule(
```

Comment:
I don't see anything here that triggers the application to be called immediately when the deadline is exceeded. Won't this delay all deadline notifications to the application by 1 second?

Reply:
All notifications to the application are sent in `startDeadlineNotifyApplicationTimer()`, which follows the initial deadline.

Reply:
In the case calloption's deadline equals to the context deadline, `startDeadlineNotifyApplicationTimer()` is not called when the ClientCall is started.